File: ICPC/2022/ICPC day2/q1.cpp
#include <iostream>
#include <stack>
#include <vector>
using namespace std;

int makeNum(const string& str)
{
	int num = 0;
	for (int i = 0; str[i]; i++)
	{
		num = num * 10 + str[i] - '0';
	}
	return num;
}

string intToString(int num)
{
	string out = "";
	do
	{
		out.push_back(num%10+'0');
		num /= 10;
	}while (num != 0);
	string rout = "";
	for (int i = 0;i< out.size(); i++)
	{
		rout.push_back(out[out.size()-1-i]);
	}
	return rout;
}

bool checkSubstr(const string& str, int start, int stop)
{
	int count = 0;
	stack<char> stk;
	for(int i=start;i<stop;i++)
	{
		if(str[i]=='(')
		{
			count++;
		}
		else
		{
			count--;
		}
	}
	if (count==0)
	{
		return true;
	}
	else
	{
		return false;
	}
} 

string testCase(const string& input)
{
	vector<int> countSubstr(10000,0);
	int maxSubstr=-1;
	stack<char> stk;
	int start;
	for (int i = 0; i < input[i]; i++)
	{
		if (input[i] == '(')
		{
			if (stk.empty())
			{
				start = i;
			}
			stk.push('(');	
		}
		else 
		{
			if(stk.empty())
			{
				continue;
			}
			else
			{
				stk.pop();
				if(stk.empty())
				countSubstr[i-start] += 1;
				if (i-start > maxSubstr)
				{
					maxSubstr = i-start;
				}
			}
		}
	}
	if (maxSubstr == -1) { return "0 1"; }
	return intToString(maxSubstr+1) + " " + intToString(countSubstr[maxSubstr])+"\n";
}

int main()
{
	int testCases;;
	cin >> testCases;
	vector<string> inputs(testCases);
	for (int i = 0; i < testCases; i++)
	{
		cin >> inputs[i];
	}
	string out="";
	for (int i = 0; i < testCases; i++)
	{
		out += testCase(inputs[i]);
	}
	cout << out;
}

File: ICPC/2022/ICPC day2/q1.py

testCases = input()
outStr = ""
inputs = []
for _ in range(int(testCases)):
    t = input()
    inputs.append(t)
for i in range(int(testCases)):
    stk = []
    line = inputs[i]
    curLen = 0
    countSubstr = {}
    maxLen = -1
    for i in range(len(line)):
        if line[i] == "(":
            stk.append("(")
        elif len(stk)==0 and line[i]==")":
            if curLen!=0:
                if curLen not in countSubstr:
                    countSubstr[curLen] = 1
                else:
                    print(countSubstr)
                    countSubstr[curLen] += 1
                if maxLen<curLen:
                    maxLen = curLen
                curLen = 0
        else:
            stk.pop()
            curLen += 2
        if len(line)==i+1 and curLen != 0 and len(stk)==0:
            countSubstr[curLen] += 1
            if maxLen<curLen:
                maxLen = curLen
            curLen = 0
    if maxLen == -1:
        print("0 1")            
    else:
        print(f"{maxLen} {countSubstr[maxLen]}")